pinot-core/src/main/java/org/apache/pinot/core/common/RowBasedBlockValueFetcher.java

```diff
@@ -0,0 +1,217 @@
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.pinot.core.common;
+
+import java.io.Serializable;
+import org.apache.pinot.common.data.FieldSpec.DataType;
+
+
+public class RowBasedBlockValueFetcher {
+  private final ValueFetcher[] _valueFetchers;
+
+  public RowBasedBlockValueFetcher(BlockValSet[] blockValSets) {
+    int numColumns = blockValSets.length;
+    _valueFetchers = new ValueFetcher[numColumns];
+    for (int i = 0; i < numColumns; i++) {
+      _valueFetchers[i] = createFetcher(blockValSets[i]);
+    }
+  }
+
+  public Serializable[] getRow(int docId) {
+    int numColumns = _valueFetchers.length;
+    Serializable[] row = new Serializable[numColumns];
+    for (int i = 0; i < numColumns; i++) {
+      row[i] = _valueFetchers[i].getValue(docId);
+    }
+    return row;
+  }
+
+  private ValueFetcher createFetcher(BlockValSet blockValSet) {
+    DataType valueType = blockValSet.getValueType();
+    if (blockValSet.isSingleValue()) {
+      switch (valueType) {
+        case INT:
+          return new IntSingleValueFetcher(blockValSet.getIntValuesSV());
+        case LONG:
+          return new LongSingleValueFetcher(blockValSet.getLongValuesSV());
+        case FLOAT:
+          return new FloatSingleValueFetcher(blockValSet.getFloatValuesSV());
+        case DOUBLE:
+          return new DoubleSingleValueFetcher(blockValSet.getDoubleValuesSV());
+        case STRING:
+          return new StringSingleValueFetcher(blockValSet.getStringValuesSV());
+        case BYTES:
+          return new BytesValueFetcher(blockValSet.getBytesValuesSV());
+        default:
+          throw new IllegalStateException();
+      }
+    } else {
+      switch (valueType) {
+        case INT:
+          return new IntMultiValueFetcher(blockValSet.getIntValuesMV());
+        case LONG:
+          return new LongMultiValueFetcher(blockValSet.getLongValuesMV());
+        case FLOAT:
+          return new FloatMultiValueFetcher(blockValSet.getFloatValuesMV());
+        case DOUBLE:
+          return new DoubleMultiValueFetcher(blockValSet.getDoubleValuesMV());
+        case STRING:
+          return new StringMultiValueFetcher(blockValSet.getStringValuesMV());
+        default:
+          throw new IllegalStateException();
```

Comment:
May be add a message with type in exception?

Reply:
Done